### PR TITLE
Attempts to optimize connections

### PR DESCRIPTION
**Pull request by jenf**
Skip reconnection walking if all new paths do not connect to anything,
simplify it to just migrate the references to the new hex

### CODE REVIEW

**Comment by jenf** [comment]
@tobymao here's an optimisaiton, a few connections related tests are failing.

The code can probably be done better, but I won't be able to look at it for a bit

**Comment by jenf** [comment]
It can probably be generalized to only connect new exits if unambigious, and migrate the existing ones.
Original:
`     TOTAL    (pct)     SAMPLES    (pct)     FRAME
     16474  (46.1%)        4866  (13.6%)     Engine::Part::Path#walk
      5803  (16.2%)        2404   (6.7%)     Engine::Connection.valid_connection?
      2311   (6.5%)        2311   (6.5%)     Engine::Tile#id
      2125   (5.9%)        2125   (5.9%)     (sweeping)
      3739  (10.5%)        1433   (4.0%)     Engine::Part::Base#id
      1332   (3.7%)        1332   (3.7%)     (marking)`

New:
`     10968  (36.0%)        3634  (11.9%)     Engine::Part::Path#walk
      1531   (5.0%)        1531   (5.0%)     (sweeping)
      3187  (10.5%)        1300   (4.3%)     Engine::Connection.valid_connection?
      1259   (4.1%)        1259   (4.1%)     Engine::Tile#id
       915   (3.0%)         915   (3.0%)     (marking)
`

File: lib/engine/hex.rb [hunk 1]
```diff
@@ -205,6 +218,29 @@ def connect!
...
+      #wanted_hex = 'D6'
+      #if id == wanted_hex
+      #  puts "old", @tile.paths 
+      #  puts "new", tile.paths
+      #  puts tile.paths.map(&:exits)
+      #  puts tile.paths.map(&:exits).uniq
+      #end
+
+      # The migration code cannot handle Chicago like upgrades
+      return false if tile.paths.map(&:exits) != tile.paths.map(&:exits).uniq
```
Comment (tobymao) [inline lib/engine/hex.rb:232]: this doesn't seem quite right and neither does just tile.exits

**Comment by tobymao** [comment]
it seems like we could actually save some time by just caching ids?

**Comment by tobymao** [comment]
just thinking aloud, would it be possible to instead of connecting every lay, to only connect stuff when we need it for running routes?

**Comment by tobymao** [comment]
Let's chat more tomorrow, but it may be possible to do away with connections as we know it. It's going to be quite a large piece of work but will probably shave off a huge amount of time.

It seems like connections are only used in routing, which means that we really only need have them when a user is acutally creating a route. When rehydrating them from the DB, we can store the literal representation so in a sense they are only calculated when needed, and only when the user is actually triggering them. That means no connections are ever created when loading games.

**Comment by jenf** [comment]
closing this as we need a better solution

**Pull request closed without merging**